### DISCUSSION TITLE
Restore image and file based clipboard contents after an expansion on macOS

### DISCUSSION THREAD

**Discussion by mark1bean** [espanso/espanso#1226] [Feature requests and Ideas]
Steps to reproduce:

Example 1:
1. In Finder, select a file and Copy
2. New Folder
3. Perform espanso text replacement (eg. "qqd" -> today's date)
4. Paste (into that folder)
Result: nothing
Expected: paste the file from 1.

Example 2:
1. In Adobe Illustrator, copy a page item
2. Double-click artboard name to edit and perform same text replacement as example 1.
3. Paste (into Illustrator document)
Result: pastes raw source of clipboard as text
Expected: pastes the item copied from 1.

MacOS Big Sur 11.5.1 (20G80), 13" M1 laptop.

**Comment by federico-terzi** [comment]
Hey @mark1bean,

I'm sorry you experienced this problem! Unfortunately, this is a current limitation of the Espanso architecture: for some matches, Espanso has to use the clipboard to perform the expansion. After the expansion, espanso tries to recover the previous content, but currently, it can only do so if it was text-based. It cannot recover images or file contents.

Have you specified `backend: Clipboard` in your `default.yml`? If that's the case, removing it would likely mitigate the problem.

Cheers :)

**Reply by mark1bean** [reply]
Thanks @federico-terzi, I hadn't specified `backend: Clipboard`. I changed it just now to `backend: Inject` but it didn't make any difference to the problem (of course I didn't expect it to). Anyway, thanks for lodging the discussion. I would LOVE this to be solved some how—it is an issue for me approximately 5-10% of the time I invoke Espanso, and I use it *a lot*! :)
